image-registry/automate-pruning.md

```diff
@@ -0,0 +1,252 @@
+pruning process.
+
+## Design Details
+
+### API
+
+Proposed changes to the image registry operator's config spec:
+
+```yaml
+spec:
+  ...
+  pruning:
+    schedule: "*/1 * * * *"
+    keepTagRevisions: 3
+    keepYoungerThan: 60m
+    resources: {}
+    affinity: {}
+    nodeSelector: {}
+    tolerations: {}
+    history:
+      successfulJobsHistoryLimit: 3
+      failedJobsHistoryLimit: 1
+```
+
+1. `schedule:` Cron formatted schedule, defaults to daily at midnight. If set to empty, job is suspended.
+2. `keepTagRevisions:` number of revisions per tag to keep. Default 3.
+3. `keepYoungerThan:` retain images younger than this duration. Default 60m.
+4. `resources:` standard Pod resource requests and limits.
+5. `affinity:` standard Pod affinity
+6. `nodeSelector:` standard Pod node selector
+7. `tolerations:` standard Pod tolerations
+8. `history:` sets the job history to be retained on the cluster
+   1. `successfulJobsHistoryLimit:` maximum number of successful jobs to retain
+   2. `failedJobsHistoryLimit:` maximum number of failed jobs to retain
+
+### Fixed Configuration
+
+The following settings for the pruning `CronJob` cannot be configured:
+
+1. `concurrencyPolicy: Forbid` - this prevents concurrent cron jobs from
+running.
+2. `suspend` - based on the provided `schedule`. If empty `suspend = true`.
+3. Pruning CLI:
+   1. `--prune-registry` - always `true` (default)
+   2. `--all` - always `true` (default)
+   3. `--prune-over-size-limit` - not set
+   4. `--confirm` - always `true` so pruning runs
+
+### Telemetry
+
+1. Expose a metric indicating if pruning has been enabled or disabled.
+2. Fire an alert if the most recent pruning run failed.
+
+### Test Plan
+
+- `e2e-*-operator` suite can verify that the `CronJob` is configured properly on the cluster.
+- A separate suite can also verify that the pruning works end to end:
+  - Create a build with imagestream tag. Run the build.
+  - Delete build & IST.
+  - Run pruning.
+  - Verify image is pruned.
+
+### Graduation Criteria
+
+#### Tech Preview
+
+At minimum this feature can be broadcast as tech preview in 4.3. Requirements
+for reaching tech preview:
+
+1. Operator installs the `CronJob` and associated RBAC.
+2. Pruner is disabled by default - cluster admins can opt-in.
+3. Tests to ensure that the `CronJob` is installed and configured properly.
+4. Telemetry which indicates the pruner has been enabled.
+
+#### Generally Available
+
+Ideally this feature is generally available in 4.3. Requirements for reaching
+GA:
+
+1. Pruner is enabled by default - cluster admins can opt-out.
+2. E2e testing of pruning via CI (this may require a separate test suite),
+voting for `image-registry` and `cluster-image-registry-operator`.
+3. Alerting if the most recent pruning run failed.
+
+### Upgrade / Downgrade Strategy
+
+On new installs and upgrades, the pruner will be installed on the cluster.
+For GA the pruner will be enabled and run daily at midnight.
```

Comment:
it's only going to be enabled automatically on new installs, not upgrades, right?

Reply:
I don't like the idea that new clusters and upgraded clusters have different default configuration. For some customers it's important to have the pruner enabled.

@bparees would you prefer to have an alert instead of enabling pruning during upgrades?

Reply:
I don't like having different defaults either, but we definitely cannot just turn pruning on on someone's cluster as part of an upgrade, we might destroy a bunch of images.

and we definitely want to have this on in general, which means turning it on by default when we can (new installs).

All things considered i don't think having some clusters w/ pruning enabled and some with it not enabled is the worst thing in the world..  since admins can always turn it off anyway, we could/will still end up in that situation regardless of how we default new installs+upgrades.

Having an alert that pruning is not enabled on the cluster seems reasonable to me, to encourage those admins to enable it.

Reply:
@bparees added separate section for upgrades.